File: commonfamily/src/main/java/com/koenig/communication/messages/UpdatesMessage.java
package com.koenig.communication.messages;


import com.koenig.commonModel.Component;
import com.koenig.commonModel.Item;
import com.koenig.commonModel.database.DatabaseItem;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by Thomas on 11.01.2017.
 */

public class UpdatesMessage<T extends Item> extends FamilyMessage {

    public static final String NAME = "UpdatesMessage";
    private List<DatabaseItem<T>> items;

    public UpdatesMessage(List<DatabaseItem<T>> items) {
        super(Component.FINANCE);
        this.items = items;
    }

    public UpdatesMessage(int version, Component component, String fromId, String toId, List<DatabaseItem<T>> items) {
        super(component);
        this.setVersion(version);
        this.setFromId(fromId);
        this.setToId(toId);
        this.items = items;
    }

    public UpdatesMessage(int version, Component component, String fromId, String toId, ByteBuffer buffer) {
        super(component);
        this.setVersion(version);
        this.setFromId(fromId);
        this.setToId(toId);
        int size = buffer.getShort();
        items = new ArrayList<>(size);
        for (int i = 0; i < size; i++) {
            items.add(new DatabaseItem<T>(buffer));
        }
    }


    public List<DatabaseItem<T>> getItems() {
        return items;
    }

    public String getName() {
        return NAME;
    }


    @Override
    protected int getContentLength() {
        int size = 2;
        for (DatabaseItem<T> item : items) {
            size += item.getByteLength();
        }
        return size;
    }

    @Override
    protected void writeContent(ByteBuffer buffer) {
        buffer.putShort((short) items.size());
        for (DatabaseItem<T> item : items) {
            item.writeBytes(buffer);
        }
    }

    @Override
    public String toString() {
        return "UpdatesMessage{" +
                "items=" + items +
                '}';
    }
}

File: commonfamily/src/main/java/com/koenig/communication/messages/UpdatesMessage.kt
package com.koenig.communication.messages


import com.koenig.commonModel.Component
import com.koenig.commonModel.Item
import com.koenig.commonModel.database.DatabaseItem
import java.nio.ByteBuffer
import java.util.*

/**
 * Created by Thomas on 11.01.2017.
 */

class UpdatesMessage<T : Item> : FamilyMessage {
    var items: MutableList<DatabaseItem<T>>

    override val name: String
        get() = NAME


    override val contentLength: Int
        get() {
            var size = 2
            for (item in items) {
                size += item.byteLength
            }
            return size
        }

    constructor(items: MutableList<DatabaseItem<T>>) : super(Component.FINANCE) {
        this.items = items
    }

    constructor(version: Int, component: Component, fromId: String, toId: String, items: MutableList<DatabaseItem<T>>) : super(component) {
        this.version = version
        this.fromId = fromId
        this.toId = toId
        this.items = items
    }

    constructor(version: Int, component: Component, fromId: String, toId: String, messageId: String, buffer: ByteBuffer) : super(component, messageId) {
        this.version = version
        this.fromId = fromId
        this.toId = toId
        val size = buffer.short.toInt()
        items = ArrayList(size)
        for (i in 0 until size) {
            items.add(DatabaseItem(buffer))
        }
    }


    override fun writeContent(buffer: ByteBuffer) {
        buffer.putShort(items.size.toShort())
        for (item in items) {
            item.writeBytes(buffer)
        }
    }

    override fun toString(): String {
        return "UpdatesMessage{" +
                "items=" + items +
                '}'.toString()
    }

    companion object {

        val NAME = "UpdatesMessage"
    }
}
